File: apps/jsx/src/mrly/life/step.ts
import { Cell2d } from "../two/models";
import type { Boundary } from "./enums";

export function step(
  grid: Cell2d,
  mask: number[][],
  birth: number[],
  survive: number[],
  boundary: Boundary,
): Cell2d {
  const neighborCell = grid.neighbors(mask, 1, boundary);
  const counts = neighborCell.tags;
  const types = grid.types;
  const h = grid.height;
  const w = grid.width;
  const birthSet = new Set(birth);
  const surviveSet = new Set(survive);
  const next: number[][] = [];
  for (let y = 0; y < h; y++) {
    const row: number[] = new Array(w);
    for (let x = 0; x < w; x++) {
      const n = counts[y][x];
      const alive = types[y][x] === 1;
      row[x] = (alive ? surviveSet.has(n) : birthSet.has(n)) ? 1 : 0;
    }
    next.push(row);
  }
  return new Cell2d({ types: next });
}

export function gridKey(grid: number[][]): string {
  const parts: string[] = [];
  for (const row of grid) parts.push(row.join(""));
  return parts.join("|");
}

export function gridsEqual(a: number[][], b: number[][]): boolean {
  if (a.length !== b.length) return false;
  for (let y = 0; y < a.length; y++) {
    const ra = a[y];
    const rb = b[y];
    if (ra.length !== rb.length) return false;
    for (let x = 0; x < ra.length; x++) if (ra[x] !== rb[x]) return false;
  }
  return true;
}

File: pkgs/rs/mrlymath/src/life/step.rs
use super::Boundary;
use crate::dim::models::counting_dtype;
use crate::two::Cell2d;
use mrlycore::cell::Cell;
use mrlycore::errors::Result;
use mrlycore::tensor::Tensor;

/// Advances a grid one generation under birth and survive counts, a neighbor mask and a boundary.
pub fn next_grid(
    cell: &Cell2d,
    birth: &[usize],
    survive: &[usize],
    mask: &Tensor,
    boundary: Boundary,
) -> Result<Cell2d> {
    let dtype = counting_dtype(mask);
    let counted = Cell::new(cell.types().clone()).neighbors(mask, 1, boundary.wrap(), dtype)?;
    let neighbors = counted.tags.expect("neighbors sets tags");
    let types = &counted.types;
    let mut next = Tensor::new(types.shape.clone());
    for (i, slot) in next.bytes_mut().iter_mut().enumerate() {
        let n = neighbors.at(i) as usize;
        let alive = types.bytes()[i] == 1;
        let lives = if alive {
            survive.contains(&n)
        } else {
            birth.contains(&n)
        };
        *slot = if lives { 1 } else { 0 };
    }
    Ok(Cell2d::new(next))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::life::moore;
    use crate::two::designs;
    #[test]
    fn conway_blinker_oscillates() {
        let mut t = Tensor::new(vec![5, 5]);
        t.set(&[1, 2], 1);
        t.set(&[2, 2], 1);
        t.set(&[3, 2], 1);
        let cell = Cell2d::new(t);
        let mask = moore().types().clone();
        let next = next_grid(&cell, &[3], &[2, 3], &mask, Boundary::Constant).unwrap();
        assert_eq!(next.types().get(&[2, 1]), 1);
        assert_eq!(next.types().get(&[2, 2]), 1);
        assert_eq!(next.types().get(&[2, 3]), 1);
        assert_eq!(next.types().get(&[1, 2]), 0);
        assert_eq!(next.types().get(&[3, 2]), 0);
        let back = next_grid(&next, &[3], &[2, 3], &mask, Boundary::Constant).unwrap();
        assert_eq!(back.types(), cell.types());
    }
    #[test]
    fn empty_stays_empty() {
        let cell = designs::zeros(3, 1).unwrap();
        let mask = moore().types().clone();
        let next = next_grid(&cell, &[3], &[2, 3], &mask, Boundary::Constant).unwrap();
        assert_eq!(next.types().sum(), 0);
    }
    #[test]
    fn wide_masks_count_beyond_a_byte() {
        use mrlycore::tensor::Tensor;
        let mut mask = Tensor::full(vec![17, 17], 1);
        mask.set(&[8, 8], 0);
        let full = 17 * 17 - 1;
        let cell = Cell2d::new(Tensor::full(vec![21, 21], 1));
        let kept = next_grid(&cell, &[], &[full], &mask, Boundary::Wrap).unwrap();
        assert_eq!(kept.types().sum() as usize, 21 * 21);
        let gone = next_grid(&cell, &[], &[full - 1], &mask, Boundary::Wrap).unwrap();
        assert_eq!(gone.types().sum(), 0);
    }
}
